ossl/src/lib.rs

```diff
@@ -0,0 +1,2378 @@
+    ///
+    /// The caller must ensure the lifetime of `val` exceeds the `OsslParam`'s
+    /// usage.
+    #[allow(dead_code)]
+    pub fn add_int(&mut self, key: &CStr, val: &'a c_int) -> Result<(), Error> {
+        if self.finalized {
+            return Err(Error::new(ErrorKind::WrapperError));
+        }
+
+        let param = unsafe {
+            OSSL_PARAM_construct_int(
+                key.as_ptr(),
+                val as *const _ as *mut c_int,
+            )
+        };
+        self.p.to_mut().push(param);
+        self.br.push(BorrowedReference::Int(val));
+        Ok(())
+    }
+
+    /// Adds an `c_uint` parameter using an owned value.
+    #[allow(dead_code)]
+    pub fn add_owned_uint(
+        &mut self,
+        key: &CStr,
+        val: c_uint,
+    ) -> Result<(), Error> {
+        if self.finalized {
+            return Err(Error::new(ErrorKind::WrapperError));
+        }
+
+        let v = val.to_ne_bytes().to_vec();
+
+        let param = unsafe {
+            OSSL_PARAM_construct_uint(
+                key.as_ptr(),
+                v.as_ptr() as *const _ as *mut c_uint,
+            )
+        };
+        self.v.push(v);
+        self.p.to_mut().push(param);
+        Ok(())
+    }
+
+    /// Adds an `c_uint` parameter using an owned value.
+    #[allow(dead_code)]
+    pub fn add_owned_int(
+        &mut self,
+        key: &CStr,
+        val: c_int,
+    ) -> Result<(), Error> {
+        if self.finalized {
+            return Err(Error::new(ErrorKind::WrapperError));
+        }
+
+        let v = val.to_ne_bytes().to_vec();
+
+        let param = unsafe {
+            OSSL_PARAM_construct_int(
+                key.as_ptr(),
+                v.as_ptr() as *const _ as *mut c_int,
+            )
+        };
+        self.v.push(v);
+        self.p.to_mut().push(param);
+        Ok(())
+    }
+
+    /// Finalizes the `OSSL_PARAM` array by adding the end marker.
+    ///
+    /// Must be called before `as_ptr` or `as_mut_ptr` can be safely used.
+    pub fn finalize(&mut self) {
+        if !self.finalized {
+            self.p.to_mut().push(unsafe { OSSL_PARAM_construct_end() });
+            self.finalized = true;
+        }
+    }
+
+    /// Returns a const pointer to the finalized `OSSL_PARAM` array.
+    ///
+    /// Panics if the array has not been finalized.
+    #[allow(dead_code)]
+    pub fn as_ptr(&self) -> *const OSSL_PARAM {
+        if !self.finalized {
+            panic!("Unfinalized OsslParam");
+        }
+        self.p.as_ref().as_ptr()
+    }
+
+    /// Returns a mutable pointer to the finalized `OSSL_PARAM` array.
+    ///
+    /// Panics if the array has not been finalized.
+    #[allow(dead_code)]
+    pub fn as_mut_ptr(&mut self) -> *mut OSSL_PARAM {
+        if !self.finalized {
+            panic!("Unfinalized OsslParam");
+        }
+        self.p.to_mut().as_mut_ptr()
+    }
+
+    unsafe fn int_mut_ptr(&self) -> *mut OSSL_PARAM {
```

Comment:
should this have doc text (even though its unsafe)?

Reply:
sure, I'll add